src/filter/TargetingFilter.ts

```diff
@@ -0,0 +1,143 @@
+                parameters.Audience.Exclusion.Users.includes(appContext.userId)) {
+                return false;
+            }
+            // check if the user is in a group within exclusion list
+            if (appContext?.groups !== undefined &&
+                parameters.Audience.Exclusion.Groups !== undefined) {
+                for (const excludedGroup of parameters.Audience.Exclusion.Groups) {
+                    if (appContext.groups.includes(excludedGroup)) {
+                        return false;
+                    }
+                }
+            }
+        }
+
+        // check if the user is being targeted directly
+        if (appContext?.userId !== undefined &&
+            parameters.Audience.Users !== undefined &&
+            parameters.Audience.Users.includes(appContext.userId)) {
+            return true;
+        }
+
+        // check if the user is in a group that is being targeted
+        if (appContext?.groups !== undefined &&
+            parameters.Audience.Groups !== undefined) {
+            for (const group of parameters.Audience.Groups) {
+                if (appContext.groups.includes(group.Name)) {
+                    const audienceContextId = constructAudienceContextId(featureName, appContext.userId, group.Name);
+                    const rolloutPercentage = group.RolloutPercentage;
+                    if (TargetingFilter.#isTargeted(audienceContextId, rolloutPercentage)) {
+                        return true;
+                    }
+                }
+            }
+        }
+
+        // check if the user is being targeted by a default rollout percentage
+        const defaultContextId = constructAudienceContextId(featureName, appContext?.userId);
+        return TargetingFilter.#isTargeted(defaultContextId, parameters.Audience.DefaultRolloutPercentage);
+    }
+
+    static #isTargeted(audienceContextId: string, rolloutPercentage: number): boolean {
+        if (rolloutPercentage === 100) {
+            return true;
+        }
+        // Cryptographic hashing algorithms ensure adequate entropy across hash values.
+        const contextMarker = stringToUint32(audienceContextId);
+        const contextPercentage = (contextMarker / 0xFFFFFFFF) * 100;
+        return contextPercentage < rolloutPercentage;
+    }
+
+    static #validateParameters(parameters: TargetingFilterParameters): void {
+        if (parameters.Audience.DefaultRolloutPercentage < 0 || parameters.Audience.DefaultRolloutPercentage > 100) {
+            throw new Error("Audience.DefaultRolloutPercentage must be a number between 0 and 100.");
+        }
+        // validate RolloutPercentage for each group
+        if (parameters.Audience.Groups !== undefined) {
+            for (const group of parameters.Audience.Groups) {
+                if (group.RolloutPercentage < 0 || group.RolloutPercentage > 100) {
+                    throw new Error(`RolloutPercentage of group ${group.Name} must be a number between 0 and 100.`);

```

Comment:
"between 0 and 100" makes it sound like it needs to be 1-99. Could do any of:

1. "must be no less than 0 and no more than 100."
2. "must be a number between 0 and 100, inclusive.
3. "is out of the accepted range." (this is what dotnet says today- but it's less clear)